scapy/contrib/isotp.py

```diff
@@ -57,100 +52,1171 @@ def answers(self, other):
+
+        if hasattr(msg, "src"):
+            msg.src = self.src
+        if hasattr(msg, "dst"):
+            msg.dst = self.dst
+        if hasattr(msg, "exsrc"):
+            msg.exsrc = self.exsrc
+        if hasattr(msg, "exdst"):
+            msg.exdst = self.exdst
+        return msg
+
+
+ISOTPSocket = ISOTPSoftSocket
+
+
+class CANReceiverThread(Thread):
+    """
+    Helper class that receives CAN frames and feeds them to the provided
+    callback. It relies on CAN frames being enqueued in the CANSocket object
+    and not being lost if they come before the sniff method is called. This is
+    true in general since sniff is usually implemented as repeated recv(), but
+    might be false in some implementation of CANSocket
+    """
+
+    def __init__(self, can_socket, callback, identifier=None):
+        """
+        Initialize the thread. In order for this thread to be able to be
+        stopped by the destructor of another object, it is important to not
+        keep a reference to the object in the callback function.
+
+        :param socket: the CANSocket upon which this class will call the
+                       sniff() method
+        :param callback: function to call whenever a CAN frame is received
+        :param identifier: the CAN identifier of the frames to accept
+        """
+        self.socket = can_socket
+        self.identifier = identifier
+        self.callback = callback
+        self.exiting = False
+
+        Thread.__init__(self)
+
+    def run(self):
+        ins = self.socket
+        did = self.identifier
+
+        while 1:
+            pkts = ins.sniff(timeout=1, count=1)
+            if len(pkts) == 1:
+                cf = pkts[0]
+            else:
+                cf = None
+            if self.exiting:
+                return
+            if cf is None:
+                continue
+            if did is not None and cf.identifier != did:
+                continue
+
+            self.callback(cf)
+
+    def stop(self):
+        self.exiting = True
+
+
+"""ISOTPSoftSocket definitions."""
+
+# Enum states
+ISOTP_IDLE = 0
+ISOTP_WAIT_FIRST_FC = 1
+ISOTP_WAIT_FC = 2
+ISOTP_WAIT_DATA = 3
+ISOTP_SENDING = 4
+
+# /* Flow Status given in FC frame */
+ISOTP_FC_CTS = 0  # /* clear to send */
+ISOTP_FC_WT = 1  # /* wait */
+ISOTP_FC_OVFLW = 2  # /* overflow */
+
+
+class ISOTPSocketImplementation:
+    """
+    Implementation of an ISOTP "state machine".
+
+    Most of the ISOTP logic was taken from
+    https://github.com/hartkopp/can-isotp/blob/master/net/can/isotp.c
+
+    This class only contains logic and timing for receiving and sending ISOTP
+    messages, but doesn't implement the actual CAN input/output.
+
+    Received CAN frames should be provided to this object using the on_recv()
+    method.
+    A provided callback function will be called every time a CAN frame should
+    be sent for both data frames and flow control (e.g. ACK) frames.
+    """
+
+    class Timer:
```

Comment:
maybe a threading.timer would suite for this job

Reply:
I can't use threading.Timer directly because I need a timer that can be started multiple times, but I could make my Timer class into a wrapper for a threading.timer object